### DISCUSSION TITLE
Is junit developed using TDD?

### DISCUSSION THREAD

**Discussion by c-schembri** [junit-team/junit-framework#3777] [General]
Are the tests developed before new features are implemented or after they're implemented? Asking because I am curious.
Thanks.

**Comment by marcphilipp** [comment]
It depends... :slightly_smiling_face: We don't prescribe _when_ tests are written but only check for extensive test coverage when reviewing PRs. The core team usually starts with a failing test when working on a bug fix or new feature so there TDD is used to some extent.

**Reply by sbrannen** [reply]
I'll second what @marcphilipp said and add that we often implement tests...

- before
- after
- and simultaneously

We effectively do all three, and as Marc said, we ensure the tests are in place before the code is pushed to `main`.

As a concrete example, take a look at the work I did for the new `@FieldSource` feature.

- I created `FieldSourceTests` following TDD in the first commit (7613b86be10766a973e403df94abef0cb881e4dd).
- And I added exhaustive tests as I worked on the feature and pushed those in a subsequent commit (79ce608c9072dd6024c35e5a77a5c98be5d74121).

**Comment by mpkorstanje** [comment]
It depends.... 😄 

Generally most of my feature PRs start with a spike to validate that the concept is viable. And only once I understand what the components and interactions involved are I'll switch to writing tests before fully developing the implementation.

And generally most of my bug fix PRs start with a reproducer in a separate project. JUnit is used in combination with other tools and frameworks, and usually when starting out it is not obvious where the problem lies. Usually I'll also test the first version of the fix against the reproducer and only then refactor the reproducer into JUnit. And that works well because JUnit facilitates integration tests with the testkit module - this reduces the gap between an external reproducer and a test in the project.
